### ISSUE TITLE
How does one access raw loss tensors from an mmdetection model?

### ISSUE THREAD

**Issue by anuragxel** [open-mmlab/mmdetection#4637]
Hi, apologies if this is a newbie question. I'm looking to access the loss tensors ("cls_loss" and "bbox_loss" tensors corresponding to an image and provided GT). I can see that the model class has a function `extract_feat` which actually returns the feature map. I was digging through the roi_head class (the `StandardRoIHead`) and I can see the function `forward_train`, but executing it throws up an error (`self.bbox_assigner` is `None`, it seems). What's the best way to access the loss tensors? Here's my code:

```
data[...] = ...
model = init_detector(config_file, checkpoint_file, device='cuda:0')
test_pipeline = Compose(model.cfg.data.test.pipeline)
data = test_pipeline(data)
input_feat = model.extract_feat(data['img'])
proposal_list = model.rpn_head.simple_test_rpn(input_feat, img_metas=data['img_metas'])
losses = model.roi_head.forward_train(input_feat, data['img_metas'], proposal_list, data["gt_bboxes"], data["gt_labels"])
```
Earlier, I was trying to access the tensor by calling `model.val_step` (and slightly modifying it to save and pass on the loss values) and also modifying the config where I changed the parameter `reduction` to `'none'` to return the whole tensor instead of summing/averaging it. That seems like the wrong way to go about it, happy to hear new directions. 

FYI, I'm trying to implement this paper (https://arxiv.org/pdf/1902.02910.pdf) which computes stats on top of loss values for every bbox prediction for a single image run through inference at different scales.  I have tried to look in the issues and the codebase too, but couldn't find a satisfactory answer till now. Thanks a lot for this great library and the continuous support!

**Comment by yhcao6** [comment]
Take `bbox_head` as an example,

https://github.com/open-mmlab/mmdetection/blob/3d5b10340581fb27b9fb070a19288439c03aaf3e/mmdet/models/roi_heads/bbox_heads/bbox_head.py#L229-L238

You can set `reduction_override` to be 'none' which will return the raw loss tensor.

**Comment by deepakksingh** [comment]
Hello @anuragxel, Could you please tell me about how to access outputs of each modules.
I mean, given a batch of images, how do i access the outputs of rpn_head, roi_head, in mmdetection.
Because by default, the model only returns the loss values dictionary.

**Comment by anuragxel** [comment]
So, from my digging, what I have understood is that the mmdetection model is a pytorch model wrapper with many extra methods and decorators. The model is constructed using components gathered via the builder pattern as defined in a provided config. The components (different backbone, necks and heads) contain the code for the forward and how to apply the loss functions. And the corresponding detector type determines what components are called in what order (Example: https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/detectors/two_stage.py). Also take a look at files in mmdet/apis to see how initialize and load models and modify config before inference/training. 

As @yhcao6 pointed out, changing the `reduction_override` within the loss function in these head components (basically the corresponding class) seems to do the trick. I'm still figuring out the details of the loss function to get the outputs in the correct format myself (which I why I have still kept the issue open, just in case), but I hope this helps. @yhcao6, please correct me if I misinterpreted/misrepresented something.

**Comment by deepakksingh** [comment]
Thanks for the promt reply. I will look into them.
I was able to implement a hook to access the tensors of various modules, but I'm still figuring out how to access the ground truth and predictions of each module.
